### PR TITLE
Automatically transform constrained parameters in SVI

### LINKED ISSUE #187: Optimize parameters with constraints in SVI
Currently, our very basic SVI implementation would require users to transform the parameters to unconstrained space before handing it to SVI. Like HMC, let us do this automatically behind the scenes or provide utilities so as to avoid having to write boiler plate code for this.

### PR DESCRIPTION
**Pull request by neerajprad**
Fixes #187.

### CODE REVIEW

File: numpyro/svi.py [hunk 1]
```diff
@@ -52,32 +58,43 @@ def init_fn(rng, model_args=(), guide_args=(), params=None):
...
+                inv_transforms[site['name']] = transform
+                params[site['name']] = transform.inv(site['value'])
 
-    def update_fn(i, opt_state, rng, model_args=(), guide_args=()):
+        def constrain_fn(inv_transforms, params):
+            return {k: inv_transforms[k](v) for k, v in params.items()}
+
+        return optim_init(params), jax.partial(constrain_fn, inv_transforms)
+
+    def update_fn(i, rng, opt_state, constrain_fn, model_args=(), guide_args=()):
```
Comment (neerajprad) [inline numpyro/svi.py]: The one disadvantage to this interface is that `constrain_fn` is the only static argument to `update_fn` and we will need to mark this as a static arg for the jit to work.

Comment (fehiepsi) [reply numpyro/svi.py]: I think we can simplify this by making `constrain_fn` a local variable of `svi`, and define it in `svi_init`.

Comment (neerajprad) [reply numpyro/svi.py]: Yes, I think it is better to remove `constrain_fn` as args here. Will make the change.

File: numpyro/svi.py [hunk 2]
```diff
@@ -52,37 +60,52 @@ def init_fn(rng, model_args=(), guide_args=(), params=None):
...
             if site['type'] == 'param':
-                params[site['name']] = site['value']
-        return optim_init(params)
+                constraint = site['kwargs'].pop('constraint', constraints.real)
+                transform = biject_to(constraint)
+                inv_transforms[site['name']] = transform
+                params[site['name']] = transform.inv(site['value'])
+
+        def transform_constrained(inv_transforms, params):
+            return {k: inv_transforms[k](v) for k, v in params.items()}
```
Comment (fehiepsi) [inline numpyro/svi.py]: can we reuse `constrain_fn` logic in `hmc_util`?

File: examples/vae.py [hunk 3]
```diff
@@ -113,7 +113,7 @@ def body_fun(i, val):
             loss_sum, rng = val
             rng, = random.split(rng, 1)
             rng, batch = binarize(rng, test_fetch(i, test_idx)[0])
-            loss = svi_eval(opt_state, rng, (batch,), (batch,)) / len(batch)
+            loss = jit(svi_eval)(rng, opt_state, (batch,), (batch,)) / len(batch)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply examples/vae.py]: Sorry, I was testing something. This shouldn't be needed.

File: numpyro/svi.py (hunk 2 above)
Comment (neerajprad) [reply numpyro/svi.py]: I have moved this to `infer_util` now, and we can put more common utilities there for prediction, etc in the future.

**Pull request merged**